File: python/muse/data_structures/linked_stack.py
class Node[T]:
    def __init__(self, element: T):
        self.data: T | None = element
        self.next: Node[T] | None = None


class LinkedStack[T]:
    def __init__(self):
        self.head = None
        self.tail = None
        self.length = 0

    def size(self) -> int:
        return self.length

    def is_empty(self) -> bool:
        return self.length == 0

    def peek(self) -> T:
        if self.length == 0:
            raise RuntimeError("[PANIC - NoSuchElement]")

        return self.tail.data

    def push(self, element: T) -> None:
        node = Node(element)

        if self.length == 0:
            self.head = node
        else:
            self.tail.next = node

        self.tail = node

        self.length += 1

    def pop(self) -> None:
        if self.length == 0:
            raise RuntimeError("[PANIC - NoSuchElement]")

        target = self.tail

        if self.length == 1:
            self.head = None
            self.tail = None
        else:
            cursor = self.head
            for _ in range(self.length - 2):
                cursor = cursor.next
            cursor.next = None
            self.tail = cursor

        target.data = None

        self.length -= 1

File: go/muse/data_structures/linked_stack/linked_stack.go
package linked_stack

import "errors"

type Node[T any] struct {
	data T
	next *Node[T]
}

type LinkedStack[T any] struct {
	head   *Node[T]
	tail   *Node[T]
	length int
}

func New[T any]() *LinkedStack[T] {
	return &LinkedStack[T]{head: nil, tail: nil, length: 0}
}

func (s *LinkedStack[T]) Size() int {
	return s.length
}

func (s *LinkedStack[T]) IsEmpty() bool {
	return s.length == 0
}

func (s *LinkedStack[T]) Peek() T {
	if s.length == 0 {
		panic(errors.New("[PANIC - NoSuchElement]"))
	}

	return s.tail.data
}

func (s *LinkedStack[T]) Push(element T) {
	node := &Node[T]{data: element, next: nil}

	if s.length == 0 {
		s.head = node
	} else {
		s.tail.next = node
	}

	s.tail = node

	s.length++
}

func (s *LinkedStack[T]) Pop() {
	if s.length == 0 {
		panic(errors.New("[PANIC - NoSuchElement]"))
	}

	target := s.tail

	if s.length == 1 {
		s.head = nil
		s.tail = nil
	} else {
		cursor := s.head
		for range s.length - 2 {
			cursor = cursor.next
		}
		cursor.next = nil
		s.tail = cursor
	}

	target.data = *new(T)

	s.length--
}
